onnxruntime/core/optimizer/qdq_transformer/selectors_actions/qdq_actions.cc

```diff
@@ -273,9 +275,198 @@
+
+    } else {
+      MlasQDQTransposeBlockwiseQuantized<MLFloat16, 4, false>(weight_src.DataAsByteSpan().data(),
+                                                              scale_src.data<MLFloat16>(),
+                                                              zp_src_ptr ? zp_src_ptr->DataAsByteSpan().data() : nullptr,
+                                                              weight_dst.data<uint8_t>(),
+                                                              scale_dst.data<MLFloat16>(),
+                                                              zp_dst_ptr ? zp_dst_ptr->data<uint8_t>() : nullptr,
+                                                              true,
+                                                              static_cast<int>(K),
+                                                              static_cast<int>(N),
+                                                              static_cast<int>(block_size),
+                                                              tp.get());
+    }
+  }
+
+  ONNX_NAMESPACE::TensorProto weight_T_tp;
+  ONNX_NAMESPACE::TensorProto scale_T_tp;
+  std::unique_ptr<ONNX_NAMESPACE::TensorProto> zp_T_tp_ptr = nullptr;
+
+  weight_dst.ToProto(weight_T_tp);
+  scale_dst.ToProto(scale_T_tp);
+  if (zp_dst_ptr) {
+    zp_T_tp_ptr = std::make_unique<ONNX_NAMESPACE::TensorProto>();
+    zp_dst_ptr->ToProto(*zp_T_tp_ptr);
+  }
+
+  auto& input_defs = replacement_node.MutableInputDefs();
+  input_defs.push_back(&graph_utils::AddInitializer(graph, weight_T_tp));
+  replacement_node.MutableInputArgsCount().push_back(1);
+  input_defs.push_back(&graph_utils::AddInitializer(graph, scale_T_tp));
+  replacement_node.MutableInputArgsCount().push_back(1);
+
+  if (zp_T_tp_ptr) {
+    input_defs.push_back(&graph_utils::AddInitializer(graph, *zp_T_tp_ptr));
+    replacement_node.MutableInputArgsCount().push_back(1);
+  }
+}
+
+Status DQMatMulReplaceWithMatMulNBits::Run(Graph& graph, const NodesToOptimize& selected_nodes) const {
+  const auto attributes = ExtraAttributes(graph, selected_nodes);
+  const auto& target = selected_nodes.Target();
+
+  // create node. we'll populate the input and output defs via moves
+  auto& replacement = graph.AddNode(target.Name(),
+                                    op_type_,
+                                    target.Description(),
+                                    {},  // input defs
+                                    {},  // output defs
+                                    &attributes,
+                                    domain_);
+
+  const auto& target_provider = target.GetExecutionProviderType();
+  replacement.SetExecutionProviderType(target_provider.empty() ? kCpuExecutionProvider : target_provider);
+
+  ORT_RETURN_IF_ERROR(MoveInputOutput(graph, selected_nodes, replacement, value_moves_, false));
+
+  AddTransposedInitializers(graph, selected_nodes, replacement);
+
+  return node_remover_.Run(graph, selected_nodes);
```

Comment:
@yufenglee  do I need to delete unused initializers here?

Reply:
Ideally, but that would require being able to ensure there are no other users of the original initializer.

Reply:
I found out that the graph.Resolve() will remove and deallocate all the unused initializers. So I guess I don't need to explicitly remove them here